flink-connector-aws/flink-connector-sqs/src/main/java/org.apache.flink.connector.sqs/sink/SqsSinkWriter.java

```diff
@@ -0,0 +1,257 @@
+    }
+
+    private void handleFullyFailedRequest(
+            Throwable err,
+            List<SendMessageBatchRequestEntry> requestEntries,
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        numRecordsOutErrorsCounter.inc(requestEntries.size());
+        boolean isFatal = SQS_EXCEPTION_HANDLER.consumeIfFatal(err, getFatalExceptionCons());
+        if (isFatal) {
+            return;
+        }
+
+        if (failOnError) {
+            getFatalExceptionCons().accept(new SqsSinkException.SqsFailFastSinkException(err));
+            return;
+        }
+
+        LOG.warn(
+                "SQS Sink failed to write and will retry {} entries to SQS first request was {}",
+                requestEntries.size(),
+                requestEntries.get(0).toString(),
+                err);
+        requestResult.accept(requestEntries);
+    }
+
+    private void handlePartiallyFailedRequest(
+            SendMessageBatchResponse response,
+            List<SendMessageBatchRequestEntry> requestEntries,
+            Consumer<List<SendMessageBatchRequestEntry>> requestResult) {
+
+        if (response.failed() != null) {
+            LOG.warn(
+                    "handlePartiallyFailedRequest: SQS Sink failed to write and will retry {} entries to SQS",
+                    response.failed().size());
+            numRecordsOutErrorsCounter.inc(response.failed().size());
+        }
+
+        if (failOnError) {
+            getFatalExceptionCons().accept(new SqsSinkException.SqsFailFastSinkException());
+            return;
+        }
+
+        if (response.failed() != null) {
+            final List<SendMessageBatchRequestEntry> failedRequestEntries =
+                    new ArrayList<>(response.failed().size());
+
+            for (final BatchResultErrorEntry failedEntry : response.failed()) {
+                final Optional<SendMessageBatchRequestEntry> retryEntry =
+                        getFailedRecord(requestEntries, failedEntry.id());
+                if (retryEntry.isPresent()) {
+                    failedRequestEntries.add(retryEntry.get());
+                }
+            }
+
+            requestResult.accept(failedRequestEntries);
+        }
+    }
+
+    private Optional<SendMessageBatchRequestEntry> getFailedRecord(
+            List<SendMessageBatchRequestEntry> requestEntries, String selectedId) {
+        for (SendMessageBatchRequestEntry entry : requestEntries) {
+            if (entry.id().equals(selectedId)) {
+                return Optional.of(entry);
+            }
+        }
+        return Optional.empty();
+    }
+
+    @Internal
+    @VisibleForTesting
+    void setSqsAsyncClient(final SqsAsyncClient sqsAsyncClient) {
+        this.sqsAsyncClient = sqsAsyncClient;
+    }
+
+    @Internal
+    @VisibleForTesting
+    void setSdkAsyncHttpClient(final SdkAsyncHttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
```

Comment:
This is an antipattern, could we instead use a `ClientProvider` as shown, so we can test it without `@VisibleForTesting` instead?
Example: 
https://github.com/apache/flink-connector-aws/blob/38aafb44d3a8200e4ff41d87e0780338f40da258/flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/sink/DynamoDbSinkWriter.java#L139

Reply:
Updated with client provider